File: benchmarks/minirocket.py
import os
from timeit import default_timer as timer
from sktime.datasets import load_UCR_UEA_dataset
from sktime.transformations.panel.rocket import MiniRocket

# Precompilation
def pre():
    dir_path = os.path.expanduser('~/.sktime_ucr')
    trainX, _ = load_UCR_UEA_dataset(dataset, split='train', return_X_y=True, return_type='numpy3D', extract_path=dir_path)
    mini = MiniRocket()
    mini.fit(trainX)
    mini.transform(trainX)
pre()

datasets = ['ACSF1', 'Adiac', 'ArrowHead', 'Beef', 'BeetleFly', 'BirdChicken', 'BME', 'Car', 'CBF', 'Chinatown', 'ChlorineConcentration', 'CinCECGTorso', 'Coffee', 'Computers', 'CricketX', 'CricketY', 'CricketZ', 'Crop', 'DiatomSizeReduction', 'DistalPhalanxOutlineAgeGroup', 'DistalPhalanxOutlineCorrect', 'DistalPhalanxTW', 'Earthquakes', 'ECG200', 'ECG5000', 'ECGFiveDays', 'ElectricDevices', 'EOGHorizontalSignal', 'EOGVerticalSignal', 'EthanolLevel', 'FaceAll', 'FaceFour', 'FacesUCR', 'FiftyWords', 'Fish', 'FordA', 'FordB', 'FreezerRegularTrain', 'FreezerSmallTrain', 'Fungi', 'GunPoint', 'GunPointAgeSpan', 'GunPointMaleVersusFemale', 'GunPointOldVersusYoung', 'Ham', 'HandOutlines', 'Haptics', 'Herring', 'HouseTwenty', 'InlineSkate', 'InsectEPGRegularTrain', 'InsectEPGSmallTrain', 'InsectWingbeatSound', 'ItalyPowerDemand', 'LargeKitchenAppliances', 'Lightning2', 'Lightning7', 'Mallat', 'Meat', 'MedicalImages', 'MiddlePhalanxOutlineAgeGroup', 'MiddlePhalanxOutlineCorrect', 'MiddlePhalanxTW', 'MixedShapesRegularTrain', 'MixedShapesSmallTrain', 'MoteStrain', 'NonInvasiveFetalECGThorax1', 'NonInvasiveFetalECGThorax2', 'OliveOil', 'OSULeaf', 'PhalangesOutlinesCorrect', 'Phoneme', 'PigAirwayPressure', 'PigArtPressure', 'PigCVP', 'Plane', 'PowerCons', 'ProximalPhalanxOutlineAgeGroup', 'ProximalPhalanxOutlineCorrect', 'ProximalPhalanxTW', 'RefrigerationDevices', 'Rock', 'ScreenType', 'SemgHandGenderCh2', 'SemgHandMovementCh2', 'SemgHandSubjectCh2', 'ShapeletSim', 'ShapesAll', 'SmallKitchenAppliances', 'SmoothSubspace', 'SonyAIBORobotSurface1', 'SonyAIBORobotSurface2', 'StarLightCurves', 'Strawberry', 'SwedishLeaf', 'Symbols', 'SyntheticControl', 'ToeSegmentation1', 'ToeSegmentation2', 'Trace', 'TwoLeadECG', 'TwoPatterns', 'UMD', 'UWaveGestureLibraryAll', 'UWaveGestureLibraryX', 'UWaveGestureLibraryY', 'UWaveGestureLibraryZ', 'Wafer', 'Wine', 'WordSynonyms', 'Worms', 'WormsTwoClass', 'Yoga']

def time(t):
    t2 = timer()
    print(f'  {t2 - t} seconds')
    return timer()

for dataset in datasets:
    print(dataset)
    t = timer()

    # Skip first timer since the datasets for Python are already filtered
    t = time(t)

    dir_path = os.path.expanduser('~/.sktime_ucr')
    trainX, trainY = load_UCR_UEA_dataset(dataset, split='train', return_X_y=True, return_type='numpy3D', extract_path=dir_path)
    testX, testY = load_UCR_UEA_dataset(dataset, split='test', return_X_y=True, return_type='numpy3D', extract_path=dir_path)
    t = time(t)

    # Create model
    mini = MiniRocket()
    t = time(t)

    # Fit model
    mini.fit(trainX)
    t = time(t)

    # Transform both trainX and testX
    tr1 = mini.transform(trainX)
    tr2 = mini.transform(testX)
    t = time(t)

    # No evaluation for MiniRocket as it's just a transformer
    print("0.0")

File: benchmarks/minirocket.jl
using MLJ
using TSC
using TSC.DataSets.Loaders
import Logging

# Disable most logs
Logging.disable_logging(Logging.Info)

# Precompilation
pre() = begin
    DataSets.list_available_datasets(UCRArchive)
    DataSets.load_dataset_metadata(UCRArchive, :Chinatown)
    trainX, trainY, _, _ = DataSets.load_dataset(UCRArchive, :Chinatown, Float32)
    trainX = DataSets.dataset_flatten_to_matrix(trainX)
    trainY = categorical(trainY, levels=unique(trainY))
    mini = MiniRocketModel()
    mach = machine(mini, (trainX, :column_based))
    fit!(mach, verbosity = 0)
    transform(mach, (trainX, :column_based))
    println("Precompilation done.")
end
pre()

# Get all datasets in the UCR Archive
datasets = DataSets.list_available_datasets(UCRArchive)

for dataset in datasets
    println(dataset)

    # Load metadata
    @time (
        train_problem_name,
        train_dimension,
        train_series_length,
        train_has_timestamps,
        train_has_missing,
        train_is_classification,
        train_class_labels
    ), (
        test_problem_name,
        test_dimension,
        test_series_length,
        test_has_timestamps,
        test_has_missing,
        test_is_classification,
        test_class_labels
    ) = DataSets.load_dataset_metadata(UCRArchive, dataset)
    
    # Skip datasets with missing values and time series of unequal length
    if train_has_missing || test_has_missing || train_series_length == 0 || train_series_length == 0
        continue
    end

    # Load dataset
    @time begin
        trainX, trainY, testX, testY = DataSets.load_dataset(UCRArchive, dataset, Float32)
        trainX =  DataSets.dataset_flatten_to_matrix(trainX)
        testX =  DataSets.dataset_flatten_to_matrix(testX)
    end

    # Create model
    @time begin
        mini = MiniRocketModel()
        mach = machine(mini, (trainX, :column_based))
    end

    # Fit model
    @time fit!(mach, verbosity = 0)

    # Transform both trainX and testX
    @time begin
        t1 = transform(mach, (trainX, :column_based))
        t2 = transform(mach, (trainX, :column_based))
    end

    # No evaluation for MiniRocket as it's just a transformer
    println("0.0")
end
